plenum/server/replica.py

```diff
@@ -185,6 +186,75 @@ def clear_below_view(self, view_no):
                 del self._batches[v]
 
 
+class ConsensusDataHelper:
+    def __init__(self, consensus_data: ConsensusSharedData):
+        self.consensus_data = consensus_data
+
+    def preprepare_batch(self, pp: PrePrepare):
+        """
+        After pp had validated, it placed into _preprepared list
+        """
+        if pp in self.consensus_data.preprepared:
+            raise LogicError('New pp cannot be stored in preprepared')
+        if pp in self.consensus_data.prepared:
+            raise LogicError('New pp cannot be stored in prepared')
+        if self.consensus_data.checkpoints and pp.ppSeqNo < self.consensus_data.last_checkpoint.seqNoEnd:
+            raise LogicError('ppSeqNo cannot be lower than last checkpoint')
+        self.consensus_data.preprepared.append(pp)
+
+    def prepare_batch(self, pp: PrePrepare):
+        """
+        After prepared certificate for pp had collected,
+        it removed from _preprepared and placed into _prepared list
+        """
+        if pp not in self.consensus_data.preprepared:
+            raise LogicError('Unprepared pp must be stored in preprepared')
+        self.consensus_data.preprepared.remove(pp)
```

Comment:
According to newly findings regarding PBFT view change we shouldn't remove batch from preprepared list when we consider it prepared

Reply:
fixed